### ISSUE TITLE
Support building from a local directory / non-Packagist source

### ISSUE THREAD

**Issue by TimWolla** [php/pie#79]
This would be useful to test if what I've written into `composer.json` is correct and whether the build actually works, before committing to my changes by pushing them into a public repository and submitting them to Packagist.

This includes both the test after the initial pie-ification and before submission to Packagist and also tests after adding or removing configure flags.

see https://phpc.social/@timwolla/113367860412703002

**Comment by asgrim** [comment]
Thanks - this should be possible once #48 is complete :+1:

**Comment by asgrim** [comment]
With the PR #83, this becomes possible, and the following steps will enable you to download/build/install source code for a PIE package that has not yet been published to Packagist:

 - Use PIE with work in progress branch `asgrim:adapt-to-use-composer-installer` (see PR #83)
 - If you have not run PIE for the target PHP before:
   - Run `pie info -v foo/bar:dev-pie` (doesn't actually matter what command really)
   - First run will not work, but will have created an empty `pie.json`
 - Locate `pie.json` for your target PHP, e.g. in `~/.config/pie/php8.3_<hash>/pie.json`
 - Update `pie.json` manually to add the repository for what you're trying to do:

Example for a GitHub repo:

```json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/foo/bar"
        }
    ]
}
```

Example for a local repo:

```json
{
    "repositories": [
        {
            "type": "path",
            "url": "/path/to/foo/bar"
        }
    ]
}
```

 - Then you can use the package as normal, (using whatever version constraint is appropriate, e.g. `foo/bar:*` for local repo, `foo/bar:dev-branch-name` for a GH repo, e.g.:
   - `bin/pie info foo/bar:dev-pie` (useful for checking the `./configure` options available in PIE)
   - `bin/pie download -v foo/bar:dev-pie`
   - `bin/pie build -v foo/bar:dev-pie --enable-some-option`
   - `bin/pie install -v foo/bar:dev-pie --enable-some-option`

Given the above possibility, maybe adding a new `pie` CLI option to combine (i.e. it would add the repository into `pie.json` for you), maybe something like:

 - `--path-repository=/path/to/foo/bar`
 - `--vcs-repository=https://github.com/foo/bar`
 
Usage examples:

 - `bin/pie install --path-repository=/path/to/foo/bar foo/bar:*`
 - `bin/pie install --vcs-repository=https://github.com/foo/bar foo/bar:*`

Since [repository configuration](https://getcomposer.org/doc/05-repositories.md) can actually get quite advanced, I'm not sure we could get a fully comprehensive syntax in the CLI for defining all possible kinds of repositories, so maybe just providing the basics would suffice? I'm open to feedback on how we actually would do this though, perhaps there is a better way :D

**Comment by macintoshplus** [comment]
The minimal structure for the package is:

```json
{
    "name": "smarty/smarty",
    "version": "3.1.7",
    "dist": {
        "url": "https://www.smarty.net/files/Smarty-3.1.7.zip",
        "type": "zip"
    }
}
```

For an extension, the `dist` key is the list of URLs of the prebuild DLL archive? or source code? or whatever?

I have this idea of JSON structure. Right?

```json
{
    "name": "smarty/smarty",
    "version": "3.1.7",
    "sources": {
        "url": "https://www.smarty.net/files/Smarty-3.1.7.zip",
        "type": "zip"
        "reference": "tags/3.1.7"
    }
    "dist": {
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86_64.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-ts-vs17-x86.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86_64.zip",
        "type": "zip"
    },{
        "url": "https://www.smarty.net/files/Smarty-3.1.7-8.4-nts-vs17-x86.zip",
        "type": "zip"
    }
}
```

**Comment by asgrim** [comment]
@macintoshplus I am not sure what your latest comment means, or how it relates to building from different Composer repositories, which is what the OP is asking about. The `sources` and `dist` keys don't exist in `composer.json`. It sounds like you are trying to solve a separate problem, so please can you open a separate ticket :)

**Comment by macintoshplus** [comment]
Right @asgrim, I have open #110.

**Comment by asgrim** [comment]
As an update to this, I'm working on a separate command to list/add/remove repositories from a target PHP in PIE. It'll be something like:

 - list `pie repository:list --with-php-config=/path/to/php-config`
 - add `pie repository:add vcs https://github.com/asgrim/example-pie-extension --with-php-config=/path/to/php-config`
 - add `pie repository:add path /path/to/your/local/extension --with-php-config=/path/to/php-config`
 - remove `pie repository:remove https://github.com/asgrim/example-pie-extension --with-php-config=/path/to/php-config`
 - remove `pie repository:remove /path/to/your/local/extension --with-php-config=/path/to/php-config`

Assuming there are no objections/issues/concerns/etc, this should make the next release :)

**Comment by macintoshplus** [comment]
> As an update to this, I'm working on a separate command to list/add/remove repositories from a target PHP in PIE. It'll be something like:
> 
> * list `pie repository:list --with-php-config=/path/to/php-config`
> * add `pie repository:add vcs https://github.com/asgrim/example-pie-extension --with-php-config=/path/to/php-config`
> * remove `pie repository:remove https://github.com/asgrim/example-pie-extension --with-php-config=/path/to/php-config`
> * add `pie repository:add path /path/to/your/local/extension --with-php-config=/path/to/php-config`
> * remove `pie repository:remove /path/to/your/local/extension --with-php-config=/path/to/php-config`
> 
> Assuming there are no objections/issues/concerns/etc, this should make the next release :)

Why the `with-php-config` option? The repository is not added into the `composer.json` project file?

**Comment by asgrim** [comment]
> Why the `with-php-config` option? The repository is not added into the `composer.json` project file?

No, PIE does not use your consuming project's `composer.json` to install extensions (yet - although it [has been discussed](https://github.com/ThePHPF/pie-design/issues/13)). It uses a `pie.json` specific to each target PHP version. This is because Composer installs dependencies for your project (in `vendor/`) as you are probably aware, but PHP extensions are global per PHP install, so we have a `pie.json` per PHP install that you have. If you have one PHP install on your machine, and `php-config` is in your `$PATH`, then you do not need to specify `--with-php-config` (this is the most common scenario). So when you use these commands, PIE is modifying `$PIE_WORKING_DIRECTORY/php8.3_<somehash>/pie.json` repositories (which is what PIE uses to manage **extension** dependencies, per "Target PHP" installation).

**Comment by macintoshplus** [comment]
Ok. I can't ask Pie to install all PHP Ext for my project based on `composer` configuration?
